pkg/cpuallocator/allocator.go

```diff
@@ -389,6 +402,526 @@ func (a *allocatorHelper) takeIdleClusters() {
+				}
+			}()
+
+			var (
+				diePartA = s.usableDieCPUCount(gA.pkg, gA.die)
+				diePartB = s.usableDieCPUCount(gB.pkg, gB.die)
+				csetA    = s.cpus[gA]
+				csetB    = s.cpus[gB]
+				full     = s.full
+				part     = s.part
+				idle     *cacheGroup
+			)
+
+			if len(s.prefer) > 0 {
+				idle = s.prefer[0]
+			}
+
+			// if we are going to use idle groups prefer other groups from the same die and pkg
+			if full > 0 && idle != nil {
+				dieIdle := s.preferDieCPUCount(idle.pkg, idle.die)
+				pkgIdle := s.preferPkgCPUCount(idle.pkg)
+
+				if gA.pkg == pkgIdle && gB.pkg != pkgIdle {
+					return -1
+				}
+				if gA.pkg != pkgIdle && gB.pkg == pkgIdle {
+					return 1
+				}
+				if gA.pkg == pkgIdle && gB.pkg == pkgIdle {
+					if gA.die == dieIdle && gB.die != pkgIdle {
+						return -1
+					}
+					if gA.die != dieIdle && gB.die == pkgIdle {
+						return 1
+					}
+					// for a tie prefer looser (bigger) group, smaller group ID
+					if gA.die == dieIdle && gB.die == pkgIdle {
+						if diff := csetA.Size() - csetB.Size(); diff != 0 {
+							return -diff
+						}
+						return gA.id - gB.id
+					}
+				}
+				// equality: both are unusable, don't need to sort them
+				return 0
+			}
+
+			// if we only have used groups, prefer tighter satisfying package and die
+			total := full + part
+
+			if diePartA >= total && diePartB < total {
+				return -1
+			}
+			if diePartA < total && diePartB >= total {
+				return 1
+			}
+			if diePartA >= total && diePartB >= total {
+				if diff := diePartA - diePartB; diff != 0 {
+					return diff
+				}
+				// for a tie prefer looser (bigger) group, smaller package, die, and group IDs
+				if diff := csetA.Size() - csetB.Size(); diff != 0 {
+					return -diff
+				}
+				if gA.pkg != gB.pkg {
+					return gA.pkg - gB.pkg
+				}
+				if gA.die != gB.die {
+					return gA.die - gB.die
+				}
+				return gA.id - gB.id
+			}
+
+			// equality: both are unusable, don't need to sort them
+			return 0
+		}
+
+		sorter = &cacheGroupSorter{
+			pick:       pickGroups,
+			sortPrefer: sortIdle,
+			sortUsable: sortUsed,
+		}
+	)
+
+	log.Debug("looking for %d CPUs (prio %s) from %s", a.cnt, a.prefer, a.from)
+
+	sorter.sortCacheGroups(a)
+
+	var (
+		preferPkgCPUs int
+		usablePkgCPUs int
+		chosenPkg     int
+
+		result = a.result
+		from   = a.from
+		cnt    = a.cnt
+	)
+
+	switch {
+	case len(sorter.prefer) > 0:
+		chosenPkg = sorter.prefer[0].pkg
+		preferPkgCPUs = sorter.preferPkgCPUCount(chosenPkg)
+		usablePkgCPUs = sorter.usablePkgCPUCount(chosenPkg)
+	case len(sorter.usable) > 0:
+		chosenPkg = sorter.usable[0].pkg
+		usablePkgCPUs = sorter.usablePkgCPUCount(chosenPkg)
+	}
+
+	if preferPkgCPUs+usablePkgCPUs < a.cnt {
+		log.Debug("=> no package can satisfy the allocation")
+		return
+	}
+
+	//
+	// take full idle cache groups, splitting up the last one if necessary
+	//
+
+	log.Debug("trying to take idle cache groups...")
+	for i, g := range sorter.prefer {
+		if cnt == 0 {
```

Comment:
`cnt <= 0` would look safer than `cnt == 0` here and on line 757.

Reply:
Updated.